scheduler/src/cook/mesos/util.clj

```diff
@@ -395,48 +459,41 @@
 ;; running and waiting jobs, so it's implemented that way to keep things simple.
 (defn get-active-jobs-by-user-and-state
   "Returns all jobs for a particular user in the specified state
-   and timeframe, without a custom executor.
-   Note that this query is not performant for completed jobs, use
-   get-completed-jobs-by-user instead."
-  [db user start end state name-filter-fn include-custom-executor? pool-name]
+   and timeframe. This query works for state waiting and running only."
+  [db user start end limit state name-filter-fn include-custom-executor? pool-name]
   (let [state-keyword (case state
                         "running" :job.state/running
                         "waiting" :job.state/waiting)
         default-pool? (pool/default-pool? pool-name)
-        pool-name' (or pool-name pool/nil-pool)
-        jobs
-        (timers/time!
-          (timers/timer ["cook-mesos" "scheduler" (str "get-" (name state) "-jobs-by-user-duration")])
-          (->> (q '[:find [?j ...]
-                    :in $ ?user ?state ?start ?end ?pool-name ?default-pool
-                    :where
-                    [?j :job/state ?state]
-                    [?j :job/user ?user]
-                    [?j :job/submit-time ?t]
-                    [(<= ?start ?t)]
-                    [(< ?t ?end)]
-                    [(cook.mesos.pool/check-pool-for-listing $ ?j :job/pool ?pool-name ?default-pool)]]
-                  db user state-keyword start end pool-name' default-pool?)
-               (map (partial d/entity db))))]
-    (cond->> jobs
-             (not include-custom-executor?) (filter #(false? (:job/custom-executor %)))
-             name-filter-fn (filter #(name-filter-fn (:job/name %)))
-             (and include-custom-executor? (= :job.state/waiting state-keyword)) (remove uncommitted?))))
+        pool-name' (or pool-name pool/nil-pool)]
+    (timers/time!
+      (timers/timer ["cook-mesos" "scheduler" (str "get-" (name state) "-jobs-by-user-duration")])
+      (->>
+        (cond->> (get-jobs-by-user-and-state-and-submit db user start end state-keyword)
+                 (not include-custom-executor?) (filter #(false? (:job/custom-executor %)))
+                 name-filter-fn (filter #(name-filter-fn (:job/name %)))
+                 (and include-custom-executor? (= :job.state/waiting state-keyword)) (remove uncommitted?))
+        (filter #(pool/check-pool-for-listing % :job/pool pool-name' default-pool?))
+        ; No need to sort. We're traversing in entity_id order, so in time order.
+        (take limit)
+        doall))))
```

Comment:
Same question here about whether we should mention in the docstring that this is returning a realized collection

Reply:
See other answer